### ISSUE TITLE
InfluxDB: New Editor to support nested functions and function parameters

### ISSUE THREAD

**Issue by antoonh** [grafana/grafana#2802]
Hi,

I've upgraded my influxdb to 0.9.4.1 and upgraded grafana to 2.5.0 pre-1. Though most of the queries I build run smooth, I now get errors when using aggregate functions such as derivative with influxdb as a backend

Seems this is due to bug https://github.com/influxdb/influxdb/issues/3718

Which aims to tighten up the syntax for aggregates: 

<quote>
>If you do the query:
>
>select derivative(value) from foo where time > now() - 4h group by time(1m)
>
>That query should fail. Derivatives with a GROUP BY time clause require an aggregate function to >be in the call to derivative, like this:
>
>select derivative(first(value)) from foo where time > now() - 4h group by time(1m)

This should be validated at parse time. The error message should indicate that they can use an aggregate function inside the call to derivative, like min, max, first, last, mean, or sum.

</quote>

The grafana 2.5.0 pre-1 build exactly replicates that error.

**Comment by torkelo** [comment]
this is because they changed derivate functions in 0.9.4.1. They  made them correctly fail when you only have them and not have an aggregation function:) 

derivate functions are not an aggregate function, so you if you have group by time you need an aggregation function. 

So until Grafana InfluxDB editor is updated to support transformation functions like derivate you will have to use the raw query mode. 

You should use a query like that:
`select derivative(mean(value), 1s) from foo where time > now() - 4h group by time(1m)`

if you want 1s rate or

**Comment by sebito91** [comment]
@torkelo, confirmed this works using 0.9.4.1 and grafana-2.1.3.

```SELECT derivative(last(value)) FROM "net_packets_recv" WHERE "host" = 'hostname' AND "interface" = 'bond0' and $timeFilter group by time($interval)```

Only issue we see (which is likely influxdb) is that interval refreshes sometimes return 0 values from the query, blanking out the data. Investigating more now, but so far we see the same results on the influxdb GUI + cli.

**Comment by sebito91** [comment]
Read @kruckenb posts from this thread: https://github.com/influxdb/influxdb/pull/3383

TL;DR -- you need to align your ```group by``` time interval with something slightly larger than your sample set...and in the grafana case with the refresh interval (if using it). For example, running below yields periodic 0 results:

```SELECT derivative(last(value)) FROM "net_packets_recv" WHERE "host" = 'hostname' AND "interface" = 'bond0' and time > now() - 10m group by time(10s)```

but upping the time grouping slightly seems to be better:

```SELECT derivative(last(value)) FROM "net_packets_recv" WHERE "host" = 'hostname' AND "interface" = 'bond0' and $timeFilter group by time(15s)```

It's really annoying to have to do this, something that needs to be addressed on the influxdb side. This post is meant mostly to "just get it working" on the grafana side, but ymmv.

**Comment by sebito91** [comment]
Temp fix in place using fill(0), see https://github.com/influxdb/influxdb/pull/3383

**Comment by sebito91** [comment]
Another follow up (hopefully final), using fill(none) actually helps avoid the astronomical numbers seen during bucket alignment upon update. 

So far, things look ok...have not seen misalignment in some time.

**Comment by torkelo** [comment]
Thinking about doing a major update to how the InfluxDB editor works. Reusing some of the same ideas in the Graphite editor. 

![image](https://cloud.githubusercontent.com/assets/10999/10170053/c4f9c262-66d3-11e5-9e21-83be5b7a0544.png)

**Comment by victorhooi** [comment]
Hmm, the workaround doesn't seem to be quite working for me.

This is my normal query using `mean`, which works:

<img width="1048" alt="screenshot 2015-09-30 22 11 01" src="https://cloud.githubusercontent.com/assets/243499/10192513/3ab79e24-67c0-11e5-8361-db16c0d7ff43.png">

```
SELECT mean("opcounters_query") AS "opcounters_query" FROM "serverstatus" WHERE "project" = 'CS-244404' AND $timeFilter GROUP BY time($interval)
```

I then add the derivative function as so:
```
SELECT derivative(mean("opcounters_query"),1s) AS "opcounters_query" FROM "serverstatus" WHERE "project" = 'CS-244404' AND $timeFilter GROUP BY time($interval)
```

However, this shows no data points:

<img width="1063" alt="screenshot 2015-09-30 22 14 09" src="https://cloud.githubusercontent.com/assets/243499/10192572/a9ee6ab6-67c0-11e5-91ac-132696b06e81.png">

If I zoom out (e.g. to 30 days, or 60 days), I get a error message "Datapoints outside the range:

<img width="1093" alt="screenshot 2015-09-30 22 14 37" src="https://cloud.githubusercontent.com/assets/243499/10192848/ae08d0ee-67c2-11e5-8c97-be6105f826ac.png">

**Comment by torkelo** [comment]
@victorhooi  as I said derivative seems to be broken in InfluxDB or many seem to struggle with it when used with group by time, this seems to be a bug in InfluxDB

**Comment by torkelo** [comment]
@victorhooi I removed your comment as it distracts from what I want this issue to be about (making the editor support InfluxDB transformation functions and aggregations functions etc), if you have issues with InfluxDB features please raise issue in the InfluxDB repo

**Comment by victorhooi** [comment]
@torkelo That's ok, I understand - I wish you'd told me first, so I could save the comment, to as the InfluxDB guys - it's all good though.

Also - does this have any bearing on this issue? https://github.com/influxdb/influxdb/pull/4149

**Comment by torkelo** [comment]
@victorhooi sorry, well it is remotely connected to that issue

**Comment by victorhooi** [comment]
@torkelo Also, out of curiosity - is the https://github.com/grafana/grafana/commits/influxdb_editor_v3 branch ready at all for people to play with? Is it in a usable state at all, or should we stay away?

Or is there not much point, because it's all going to get merged to master soon anyhow? =).

**Comment by torkelo** [comment]
No it's not usable. Not sure if it will make it into 2.5 release even as there is a lot more work that needs to be done and I really should be working on CloudWatch integration instead (and 2.5 release prep)

**Comment by victorhooi** [comment]
I just noticed a recent commit to InfluxDB today:

https://github.com/influxdb/influxdb/pull/4292

This seems to have fixed the derivative issue for me - previously I was getting results like this - just a single zero:

```
> SELECT derivative(mean("opcounters_update"),1s) AS "opcounters_update" FROM "serverstatus" WHERE "project" = 'CS-244404' AND time > 1443582574s and time < 1443585233s GROUP BY time(1s)
name: serverstatus
------------------
time                    opcounters_update
1443582574000000000     0
```

However, after upgrading InfluxDB I am now getting results as expected (i.e. the derivative per second), and also in Grafana as expected:
![screenshot 2015-10-02 12 17 40](https://cloud.githubusercontent.com/assets/243499/10238198/cab21514-68ff-11e5-910c-3b97ba45cfa5.png)

I had to tweak the query in Grafana slightly - simply adding `derivative(X, 1x)` didn't work - I had to override the default GROUP BY interval (at that zoom range Grafana was setting $interval to 100ms, but I set it to 1s, and it worked then).

My question is - now that InfluxDB seems to have fixed things - is this something you can in easily to Grafana, to support just putting `derivative(x, 1s)` (or whatever time you think is suitable) around things  to at least get things working now, or does this still needs the full editor re-write?

**Comment by jwilder** [comment]
If you don't specify a second parameter to derivative, it will default to the group by time interval.  If the is no group by, it defaults to 1s.

**Comment by victorhooi** [comment]
@torkelo I'm hoping the below is useful to debugging InfluxDB queries for Grafana - but please let me know if it's not, and I can edit it out.

@jwilder Aha, sweet. Thanks for clarifying. However, I'm still a little confused.

In my case, this particular series has one data-point per second. Here is a normal `mean()` at 1-second:
```
> SELECT mean("opcounters_update") AS "opcounters_update" FROM "serverstatus" WHERE "project" = 'CS-24404' AND time > 1443583284s and time < 1443583287s GROUP BY time(1s)
name: serverstatus
------------------
time                    opcounters_update
1443583284000000000     1.8106002e+07
1443583285000000000     1.8106598e+07
1443583286000000000     1.8107324e+07
```

If I do a `mean()` with a `GROUP BY` of 100ms (which Grafana will do automatically as you zoom in), I get:
```
> SELECT mean("opcounters_update") AS "opcounters_update" FROM "serverstatus" WHERE "project" = 'CS-24404' AND time > 1443583284s and time < 1443583287s GROUP BY time(100ms)
name: serverstatus
------------------
time                    opcounters_update
1443583284000000000
1443583284100000000
1443583284200000000
1443583284300000000
1443583284400000000     1.8106002e+07
1443583284500000000
1443583284600000000
1443583284700000000
1443583284800000000
1443583284900000000
1443583285000000000
1443583285100000000
1443583285200000000
1443583285300000000
1443583285400000000     1.8106598e+07
1443583285500000000
1443583285600000000
1443583285700000000
1443583285800000000
1443583285900000000
1443583286000000000
1443583286100000000
1443583286200000000
1443583286300000000
1443583286400000000     1.8107324e+07
1443583286500000000
1443583286600000000
1443583286700000000
1443583286800000000
1443583286900000000
```
Are those null values in between the real values? Is it normal that it should be returning those?

If I do a derivative at 1s, it all works fine:
```
> SELECT derivative(mean("opcounters_update")) AS "opcounters_update" FROM "serverstatus" WHERE "project" = 'CS-24404' AND time > 1443583284s and time < 1443583287s GROUP BY time(1s)
name: serverstatus
------------------
time                    opcounters_update
1443583285000000000     596
1443583286000000000     726
```
However, once I shrink the `GROUP BY` interval to 100ms, I get nothing:
```
> SELECT derivative(mean("opcounters_update")) AS "opcounters_update" FROM "serverstatus" WHERE "project" = 'CS-24404' AND time > 1443583284s and time < 1443583287s GROUP BY time(100ms)
name: serverstatus
------------------
time                    opcounters_update
1443583284100000000
1443583284200000000
1443583284300000000
1443583284400000000
1443583284500000000
1443583284600000000
1443583284700000000
1443583284800000000
1443583284900000000
1443583285000000000
1443583285100000000
1443583285200000000
1443583285300000000
1443583285400000000
1443583285500000000
1443583285600000000
1443583285700000000
1443583285800000000
1443583285900000000
1443583286000000000
1443583286100000000
1443583286200000000
1443583286300000000
1443583286400000000
1443583286500000000
1443583286600000000
1443583286700000000
1443583286800000000
1443583286900000000
```
Is something wrong here? Or I misunderstanding how it should work?

**Comment by jwilder** [comment]
@victorhooi Since your data is stored at `1s` granularity, you are likely getting `nils` in the buckets because they are spaced `100ms` apart.  InfluxDB is querying each of those `100ms` time boundaries and calculating the mean value.  The nils indicate that there was no data between those two time points and it does not interpolate data points store at a higher precision right now.   The only thing odd to me there is that I would have expected your first query to line up the points on the seconds boundaries (`1443583284000000000, `1443583285000000000`, ...) but it seemed to pick the `400ms` boundary as the start bucket when you ran it so you must have some millisecond precision data stored on the `400ms` buckets.

The second query works because the bucket interval is calculating the mean between each second and there are points between the second boundaries.  You could run `SELECT derivative(mean("opcounters_update"), 100ms)...` to convert the the results to 100ms rate.

The last query is returning nils because of the last PR I merged.  The problem is that we can't calculate the derivative if one of the values is nil.  Before we'd just drop the row and you'd get no data which is not what people expected.  Now it just returns `nil` but you get a row.  I did have a version that just returned the value of the non-nil point but that would cause big spikes on the edge of the graph sometimes.  It is basically assuming `fill(0)` in that case which might be problematic in some cases.

I think in your case, you could probably add `fill(previous)` to the query to so that the nils are replaced with the previous values before the derivative is run.

**Comment by victorhooi** [comment]
@jwilder Ok, still parsing your comment, and making sense of it =). I think I sort of get it.

Also, I noted you said "does not interpolate data points store at a higher precision right now" - are you referring to something like:

https://github.com/influxdb/influxdb/issues/3273

It sounds like this might fix some of the confusion around aggregations/derivatives, where the requested interval is smaller than the actual data points - and it might make some of the graphing in Grafana less confusing, particularly when you try to zoom in.

Do you know how far something like that is on the InfluxDB roadmap? Or if I'm off-base, and it's not the panacea I think it is, please let me know.

**Comment by jwilder** [comment]
@victorhooi I think we should move this discussion to the influxdb issue list since it's somewhat off-topic now.

**Comment by victorhooi** [comment]
So the new InfluxDB editor will be awesome, I'm sure =).

However, as an interim step - is there any possibility of getting derivatives to work again with the existing editor?

**Comment by torkelo** [comment]
Been working on the new editor. It is starting to work.

![influxdb_v3](https://cloud.githubusercontent.com/assets/10999/11398645/987c7a50-9382-11e5-999d-8b92887995ba.gif)

**Comment by victorhooi** [comment]
That is awesome stuff! =) Can't wait to try it out when it hits `master`!

**Comment by pasancario** [comment]
@torkelo  This is a really cool. Waiting for first master push to try it. And waiting for Influxdb team to update 0.9.5 to support derivate function with max,min,last aggregation queries.

:+1:

**Comment by torkelo** [comment]
this is now merged to master and is available in the latest master. 

I added all implemented functions listed here https://influxdb.com/docs/v0.9/query_language/functions.html

They group functions into these categories,
* Aggregations	
* Selectors	
* Transformations  

Grafana does the same. What is sad however that these function categories are super bad, you cannot draw any conclusion to how and where to use a function from the categories. For example under transformations there is `STTDEV` but this does not work as a transformation function but as an aggregation function (meaning it requires a field param). 

Some functions under `Selectors` appear to work as aggregation functions (that is they work with group by time), other functions under Selectors does not seem to work when using group by time.

**Comment by torkelo** [comment]
you can try it here:
http://grafana.org/download/builds.html

Be aware that the new editor requires an update query model, so old dashboards will be updated so that queries conform to the new query model. That means that if you try 2.6 pre release builds and save a dashboard, then downgrade to Grafana 2.5 that dashboard will not work any more.

**Comment by syepes** [comment]
Great work!

**Comment by torkelo** [comment]
Any feedback from people who have tested it?

**Comment by antoonh** [comment]
Should have time to upgrade/test over the weekend.

**Comment by dswarbrick** [comment]
Having a play with it now, converting a bunch of handwritten derivative queries. Seems to work fine. Thanks for all your hard work!

**Comment by torkelo** [comment]
@dswarbrick that is great to hear! thanks for testing it and letting me know your results :)

**Comment by antoonh** [comment]
upgraded to 2.6.0-beta1, with influxdb 0.9.4.1. This definitely breaks the query editor. Can't create where clauses.

**Comment by torkelo** [comment]
@antoonh in what way can you not create where clause? that part of the editor has not changed, its the same as before.

http://play.grafana.org/dashboard/db/influxdb?panelId=1&fullscreen&edit

**Comment by antoonh** [comment]
I'm creating a query from scratch, aiming for the interface_rx data, as I'm eager to test derivatives

When I click the + after the WHERE I see a dropdown with HOST, INSTANCE, TYPE. 
When selecting that, I'd expect to be able to enter a value and so on. That doesn't happen. It reverts to just the + after the WHERE.

Behaviour seems erratic, as upon trying it multiple times (to catch a screenshot) it works every now and then. I'm seeing this with the play.grafana env too.

**Comment by torkelo** [comment]
@antoonh what browser & os are you using? Cannot replicate it. Maybe some timing issue. Are you on a slow laptop?

**Comment by torkelo** [comment]
hm.. I can replicate it if I click and hold long at the dropdown selection, but if click normally it works everytime

**Comment by antoonh** [comment]
That'll be it. main workstation has a wireless kbd that's having connection issues. Just tested from another  system and it works flawlessly. Upgraded to influxdb 0.9.5.1 as well. Will do functional testing now.

**Comment by antoonh** [comment]
Functionally: works like a charm. I get your point about the incomprehensible function categories though, but that's really for influxdb to address I'd say?

**Comment by torkelo** [comment]
yes

**Comment by antoonh** [comment]
What would be a nice to have is to have is a way to verbatim see the actual influxdb query, for easy copy&paste into influxdb cmd line, especially with the fairly arbitrary logic of influxdb's function categories.

**Comment by discoduck2x** [comment]
@torkelo  Im unable to use MAX for derivative. Sure it gives me data but as i zoom in or out the max value changes (zoom in=closer to actual value, zoom out=lower and lowee, looks more like mean)
Data is snmp in/out octets of interface so it is a growing nr bur i want the max value for each timebucket. 

Sorry no example q now as im on mobile, can add tomorrow

**Comment by torkelo** [comment]
You should report that to InfluxDB

**Comment by Deshke** [comment]
hey,...

i'm using grafana and influxdb nightlies... 

> SELECT derivative(last(value)) FROM "interface_rx" WHERE "host" = '$host' AND "instance" = 'eth0' AND "type" = 'if_octets' AND $timeFilter GROUP BY time(10s) 

works, but the auto created query (within the fancy 2.6.0-beta2 editor) returns this

> SELECT derivative("value", 10s) FROM "interface_rx" WHERE "host" = '$host' AND "instance" = 'eth0' AND "type" = 'if_octets' AND $timeFilter GROUP BY time(10s)

**Comment by torkelo** [comment]
You need to add last function

**Comment by Deshke** [comment]
crazy shit, it works :)

**Comment by antoonh** [comment]
playing a littlebit more, I've found that sometimes a typo in the influxdb query is generated from the influxdb query builder:

this query works:

`SELECT non_negative_derivative(mean("volume"), 10s) FROM "euronext" WHERE "indexname" = 'AEX' AND "entry" = 'Fonds' AND "name" = 'ASML' AND $timeFilter GROUP BY time(5m) fill(previous)
`

but when I set the derivative or non_negative_derivative time to 5min (from the dropdown), I get the following error:

Here's the query that generates the error:

`SELECT non_negative_derivative(mean("volume"), 5min) FROM "euronext" WHERE "indexname" = 'AEX' AND "entry" = 'Fonds' AND "name" = 'ASML' AND time > now() - 6h GROUP BY time(5m) fill(previous) 
`

and this is the eror coming back (from influxdb I presume):
`{
  "error": "error parsing query: found IN, expected ) at line 1, char 50",
  "message": "error parsing query: found IN, expected ) at line 1, char 50"
}`

I presume that this is due to grafana passing in "5min" verbatim, which likely should be 5m (as suggested by the other values 1m, 10m and so on in the drop down.

**Comment by torkelo** [comment]
@antoonh thanks for informing me, fixed in master

**Comment by antoonh** [comment]
Great!
